torch/lib/c10d/TcpStore.cpp

```diff
@@ -0,0 +1,274 @@
+    } else {
+      for (auto& key : keys) {
+        waiting_[key].push_back(rank);
+      }
+      keysAwaited_[rank] = keys.size();
+      tcputil::sendValue<QueryType>(socket, QueryType::KEEP_WAITING);
+    }
+  } else {
+    throw std::runtime_error("expected a query type");
+  }
+}
+
+bool TcpStoreDaemon::checkAndUpdate(std::vector<std::string>& keys) const {
+  bool ret = true;
+  for (auto it = keys.begin(); it != keys.end();) {
+    if (tcpStore_.count(*it) == 0) {
+      ret = false;
+      it++;
+    } else {
+      it = keys.erase(it);
+    }
+  }
+  return ret;
+}
+
+// TcpStore class methods
+
+TcpStore::TcpStore(const std::string& masterAddr,
+                   PortType masterPort,
+                   bool isServer)
+ : isServer_(isServer)
+ , tcpStoreAddr_(masterAddr)
+ , tcpStorePort_(masterPort)
+
+{
+  if (isServer_) {
+    // Openning up the listening socket
+    std::tie(masterListenSocket_, std::ignore) = tcputil::listen(masterPort);
+    // Now start the daemon
+    tcpStoreDaemon_ = std::unique_ptr<TcpStoreDaemon>(
+        new TcpStoreDaemon(masterListenSocket_)
+    );
+  }
+  // Connect to the daemon
+  storeSocket_ = tcputil::connect(tcpStoreAddr_, tcpStorePort_);
+}
+
+TcpStore::~TcpStore() {
+  ::close(storeSocket_);
+  if (isServer_) {
+    ::close(masterListenSocket_);
+    /**
+     * Store daemon should end because of closed connection.
+     * daemon destructor should join the thread
+     */
+    tcpStoreDaemon_.reset(nullptr);
+  }
+}
+
+void TcpStore::set(const std::string& key, const std::vector<uint8_t>& data) {
+  tcputil::sendValue<QueryType>(storeSocket_, QueryType::SET);
+  tcputil::sendString(storeSocket_, key, true);
+  tcputil::sendVector<uint8_t>(storeSocket_, data);
+}
+
+std::vector<uint8_t> TcpStore::get(const std::string& key) {
+  wait({key});
+  tcputil::sendValue<QueryType>(storeSocket_, QueryType::GET);
+  tcputil::sendString(storeSocket_, key);
+  return tcputil::recvVector<uint8_t>(storeSocket_);
+}
+
+int64_t TcpStore::add(const std::string& key, int64_t value) {
+  tcputil::sendValue<QueryType>(storeSocket_, QueryType::ADD);
+  tcputil::sendString(storeSocket_, key, true);
+  tcputil::sendValue<int64_t>(storeSocket_, value);
+  return tcputil::recvValue<int64_t>(storeSocket_);
+}
+
+bool TcpStore::check(const std::vector<std::string>& keys) {
+
+  tcputil::sendValue<QueryType>(storeSocket_, QueryType::CHECK);
+  SizeType nkeys = keys.size();
+  tcputil::sendBytes<SizeType>(storeSocket_, &nkeys, 1, (nkeys > 0));
+  for (size_t i = 0; i < nkeys; i++) {
+    tcputil::sendString(storeSocket_, keys[i], (i != (nkeys - 1)));
+  }
+  auto checkResponse = tcputil::recvValue<QueryType>(storeSocket_);
+  if (checkResponse == QueryType::STOP_WAITING) {
+    return true;
+  } else if (checkResponse == QueryType::KEEP_WAITING) {
+    return false;
+  } else {
+    throw std::runtime_error("stop_waiting or keep_waiting response expected");
+  }
```

Comment:
The server will return another QueryType::STOP_WAITING when it is done waiting. This will be unexpected by whatever function is called next. I think it would be easier to split "check" and "wait" into two paths, one that is blocking and one that is non blocking.